content/en/blog/_posts/2022-09-07-iptables-chains.md

```diff
@@ -0,0 +1,184 @@
+particular, using `KUBE-MARK-MASQ` to mark packets as needing to be
+masqueraded).
+
+In Kubernetes 1.25 SIG Network is declaring explicitly that (with one
+exception), the iptables chains that Kubernetes creates are intended
+only for Kubernetes’s own internal use, and third-party components
+should not assume that Kubernetes will create any specific iptables
+chains, or that those chains will contain any specific rules if they
+do exist.
+
+Then, in future releases, as part of [KEP-3178], we will begin phasing
+out certain chains that Kubernetes itself no longer needs. Components
+outside of Kubernetes itself that make use of `KUBE-MARK-MASQ`,
+`KUBE-MARK-DROP`, or other Kubernetes-generated iptables chains should
+start migrating away from them now.
+
+[KEP-3178]: https://github.com/kubernetes/enhancements/issues/3178
+
+## Background
+
+In addition to various service-specific iptables chains, kube-proxy
+creates certain general-purpose iptables chains that it uses as part
+of service proxying. In the past, kubelet also used iptables for a few
+features (such as setting up `hostPort` mapping for pods) and so it
+also redundantly created some of the same chains.
+
+However, with [the removal of dockershim] in Kubernetes in 1.24,
+kubelet now no longer ever uses any iptables rules for its own
+purposes; the things that it used to use iptables for are now always
+the responsibility of the container runtime or the network plugin, and
+there is no reason for kubelet to be creating any iptables rules.
+
+Meanwhile, although `iptables` is still the default kube-proxy backend
+on Linux, it is unlikely to remain the default forever, since the
+associated command-line tools and kernel APIs are essentially
+deprecated, and no longer receiving improvements. (RHEL 9 [logs a
+warning] if you use the iptables API, even via `iptables-nft`.)
+
+Although as of Kubernetes 1.25 iptables kube-proxy remains popular,
+and kubelet continues to create the iptables rules that it
+historically created (despite no longer _using_ them), third party
+software cannot assume that core Kubernetes components will keep
+creating these rules in the future.
+
+[the removal of dockershim]: https://kubernetes.io/blog/2022/02/17/dockershim-faq/
+[logs a warning]: https://access.redhat.com/solutions/6739041
+
+## Upcoming changes
+
+Starting a few releases from now, kubelet will no longer create the
+following iptables chains in the `nat` table:
+
+  - `KUBE-MARK-DROP`
+  - `KUBE-MARK-MASQ`
+  - `KUBE-POSTROUTING`
+
+Additionally, the `KUBE-FIREWALL` chain in the `filter` table will no
+longer have the functionality currently associated with
+`KUBE-MARK-DROP` (and it may eventually go away entirely).
+
+This change will be phased in via the `IPTablesOwnershipCleanup` feature gate.
+If all goes according to schedule, then the feature will become beta
+(enabled-by-default, but still able to be disabled) in Kubernetes
+1.27, and would be GA in Kubernetes 1.28.
```

Comment:
We [avoid statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future) in docs. For this article, statements about the future are OK - please consider rewording to be explicitly _less confident_ about that timeline.

(It's OK to be justifiably confident; for example, if we'd delay the v1.27 release to make sure that iptables ownership cleanup went to beta, then we can reword accordingly)

Reply:
We're not at all confident; that's what I meant by "If all goes according to schedule". 

> for example, if we'd delay the v1.27 release to make sure that iptables ownership cleanup went to beta

no, it's the exact opposite; the KEP explicitly says we will delay graduating from alpha to beta if needed. ("For example, Beta might be delayed if a commonly-used third-party component breaks badly when the feature gate is enabled")

The point of this paragraph was to give third-party developers a hint about the timeframe they should be thinking about, if this change is going to affect them. ie, if you need to make the changes described below, you should try to make them before 1.27.

I rewrote this but I'm happy to rewrite more...